src/coreclr/vm/assembly.cpp

```diff
@@ -598,29 +603,55 @@ Assembly *Assembly::CreateDynamic(AppDomain *pDomain, CreateDynamicAssemblyArgs
     }
 
     NewHolder<DomainAssembly> pDomainAssembly;
+    BOOL                      usesAllocatorFromBinderContext = FALSE;
 
     {
         GCX_PREEMP();
 
+        AssemblyLoaderAllocator* pAssemblyLoaderAllocator = nullptr;
+
+        if (pBinderContext != nullptr)
+        {
+            pBinderContext->GetLoaderAllocator((LPVOID*)&pAssemblyLoaderAllocator);
+            usesAllocatorFromBinderContext = TRUE;
+        }
+
         // Create a new LoaderAllocator if appropriate
         if ((args->access & ASSEMBLY_ACCESS_COLLECT) != 0)
         {
-            AssemblyLoaderAllocator *pAssemblyLoaderAllocator = new AssemblyLoaderAllocator();
-            pAssemblyLoaderAllocator->SetCollectible();
-            pLoaderAllocator = pAssemblyLoaderAllocator;
-
-            // Some of the initialization functions are not virtual. Call through the derived class
-            // to prevent calling the base class version.
-            pAssemblyLoaderAllocator->Init(pDomain);
+            if (pAssemblyLoaderAllocator == nullptr || !pAssemblyLoaderAllocator->IsCollectible())
```

Comment:
I wonder if we should rather fail early up the call chain if the passed in AssemblyLoadContext is not collectible. This silent fallback to creating an AssemblyLoaderAllocator when non-collectible AssemblyLoadContext was passed in might be confusing.

Reply:
I don't think that's an option. The fact that I'm running in a secondary ALC should not disable usage of dynamic assemblies. If the ALC is not collectible I would probably also still expect the dynamic assembly to be unloadable.

I'm actually leaning the other way, that is: Is it OK to "reuse" an unloadable ALC's loader allocator for dynamic assemblies? It then ties the lifetime of the unloadable ALC to the lifetime of dynamic assemblies. That might not be desirable. In general I would expect dynamic assemblies to be shorter lived?

Maybe we should keep the code as it was before - always create a new allocator if unloadability is wanted. And only change the ALC associated with it.

Reply:
Maybe it should look something like this?
```cpp
        AssemblyLoaderAllocator* pAssemblyLoaderAllocator = nullptr;

        // Create a new LoaderAllocator if appropriate
        if ((args->access & ASSEMBLY_ACCESS_COLLECT) != 0)
        {
            pAssemblyLoaderAllocator = new AssemblyLoaderAllocator();
            pAssemblyLoaderAllocator->SetCollectible();

            // Some of the initialization functions are not virtual. Call through the derived class
            // to prevent calling the base class version.
            pAssemblyLoaderAllocator->Init(pDomain);

            // Setup the managed proxy now, but do not actually transfer ownership to it.
            // Once everything is setup and nothing can fail anymore, the ownership will be
            // atomically transfered by call to LoaderAllocator::ActivateManagedTracking().
            pAssemblyLoaderAllocator->SetupManagedTracking(&args->loaderAllocator);
            createdNewAssemblyLoaderAllocator = TRUE;

            pLoaderAllocator = pAssemblyLoaderAllocator;
        }
        else
        {
            if (pBinderContext != nullptr)
            {
                pBinderContext->GetLoaderAllocator((LPVOID*)&pAssemblyLoaderAllocator);
            }

            pLoaderAllocator = pAssemblyLoaderAllocator == nullptr ? pDomain->GetLoaderAllocator() : pAssemblyLoaderAllocator;
        }
```

Reply:
Yes, it works.

Reply:
I like that better - but still would like to hear a second opinion as I'm definitely no expert on dynamic assemblies and such.